File: src/exprsndns/storage.py
"""Thread-safe JSON-backed storage for Exprsn DNS records."""

from __future__ import annotations

import json
import os
import tempfile
import threading
from pathlib import Path
from typing import Iterator

from .models import Record, RecordError, normalize_token


class Storage:
    """Append/update/delete .exprsn records persisted to a JSON file."""

    def __init__(self, path: str | os.PathLike[str]) -> None:
        self._path = Path(path)
        self._lock = threading.RLock()
        self._records: dict[str, Record] = {}
        self._load()

    @property
    def path(self) -> Path:
        return self._path

    def _load(self) -> None:
        if not self._path.exists():
            return
        with self._path.open("r", encoding="utf-8") as fh:
            raw = json.load(fh)
        for entry in raw.get("records", []):
            record = Record.from_dict(entry)
            self._records[record.token] = record

    def _flush_locked(self) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        payload = {"records": [r.to_dict() for r in self._records.values()]}
        fd, tmp = tempfile.mkstemp(
            prefix=".exprsndns-", suffix=".json", dir=str(self._path.parent)
        )
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as fh:
                json.dump(payload, fh, indent=2, sort_keys=True)
            os.replace(tmp, self._path)
        except Exception:
            if os.path.exists(tmp):
                os.unlink(tmp)
            raise

    def get(self, token: str) -> Record | None:
        key = normalize_token(token)
        with self._lock:
            return self._records.get(key)

    def list(self) -> list[Record]:
        with self._lock:
            return list(self._records.values())

    def __iter__(self) -> Iterator[Record]:
        return iter(self.list())

    def __len__(self) -> int:
        with self._lock:
            return len(self._records)

    def upsert(self, record: Record) -> Record:
        with self._lock:
            existing = self._records.get(record.token)
            if existing is not None:
                record.created_at = existing.created_at
            record.touch()
            self._records[record.token] = record
            self._flush_locked()
            return record

    def create(self, record: Record) -> Record:
        with self._lock:
            if record.token in self._records:
                raise RecordError(f"token already registered: {record.token}")
            self._records[record.token] = record
            self._flush_locked()
            return record

    def delete(self, token: str) -> bool:
        key = normalize_token(token)
        with self._lock:
            if key not in self._records:
                return False
            del self._records[key]
            self._flush_locked()
            return True

File: src/storage.js
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import { Record, RecordError, normalizeToken } from "./models.js";

export class Storage {
  constructor(filePath) {
    this.path = path.resolve(filePath);
    this._records = new Map();
    this._writing = Promise.resolve();
    this._load();
  }

  _load() {
    if (!fs.existsSync(this.path)) return;
    const raw = JSON.parse(fs.readFileSync(this.path, "utf8"));
    for (const entry of raw.records ?? []) {
      const record = Record.fromJSON(entry);
      this._records.set(record.token, record);
    }
  }

  async _flush() {
    const payload = {
      records: [...this._records.values()]
        .sort((a, b) => a.token.localeCompare(b.token))
        .map((r) => r.toJSON()),
    };
    const dir = path.dirname(this.path);
    await fsp.mkdir(dir, { recursive: true });
    const tmp = path.join(dir, `.exprsndns-${process.pid}-${Date.now()}.tmp`);
    await fsp.writeFile(tmp, JSON.stringify(payload, null, 2), "utf8");
    await fsp.rename(tmp, this.path);
  }

  _serialize(work) {
    const next = this._writing.then(work, work);
    this._writing = next.catch(() => {});
    return next;
  }

  get(token) {
    return this._records.get(normalizeToken(token)) ?? null;
  }

  list() {
    return [...this._records.values()];
  }

  get size() {
    return this._records.size;
  }

  async create(record) {
    if (this._records.has(record.token)) {
      throw new RecordError(`token already registered: ${record.token}`);
    }
    this._records.set(record.token, record);
    await this._serialize(() => this._flush());
    return record;
  }

  async upsert(record) {
    const existing = this._records.get(record.token);
    if (existing) {
      record.createdAt = existing.createdAt;
      record.ownerId = record.ownerId ?? existing.ownerId;
      record.certificateId = record.certificateId ?? existing.certificateId;
    }
    record.touch();
    this._records.set(record.token, record);
    await this._serialize(() => this._flush());
    return record;
  }

  async delete(token) {
    const key = normalizeToken(token);
    if (!this._records.has(key)) return false;
    this._records.delete(key);
    await this._serialize(() => this._flush());
    return true;
  }
}
